server.js

```diff
@@ -1,11 +1,148 @@
 'use strict';
 
-const { ApolloServer } = require('apollo-server');
+require('dotenv').config();
+
+const express = require('express');
+const bodyParser = require('body-parser');
+const session = require('express-session');
+const passport = require('passport');
+const saml = require('passport-saml');
+const { ApolloServer } = require('apollo-server-express');
+const { Op } = require('sequelize');
+const db = require('./models');
 const { schema } = require('./graphql');
 
-const server = new ApolloServer({ schema });
+const PORT = 4000;
+const HOST = `http://localhost:${PORT}`;
+// TODO: Change this session secret
+const SESSION_SECRET = 'notsecure';
+
+// This is used to configure login/logout functionality, e.g. the entry point to our IDP, 
+// the login callback, etc.
+const samlStrategy = new saml.Strategy(
+  {
+    path: '/login/callback',
+    entryPoint: `https://uwbp-paramedics.us.auth0.com/samlp/${process.env.AUTH0_CLIENT_ID}`,
+    acceptedClockSkewMs: -1, // SAML assertion not yet valid fix
+    logoutUrl: 'https://uwbp-paramedics.us.auth0.com/v2/logout',
+    additionalLogoutParams: {
+      client_id: process.env.AUTH0_CLIENT_ID,
+      federated: '',
+      returnTo: 'http://localhost:4000/logout/callback',
+    },
+  },
+  (profile, done) => {
+    // TODO: This will need to be changed when we switch to ADFS
+    const email =
+      profile[
+        'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress'
+      ];
+    db.user
+      .findOne({
+        where: { email: { [Op.like]: email } },
+      })
+      .then((matchingUser) => {
+        // Save the nameId and nameIDFormat for logout
+        /* eslint-disable-next-line no-param-reassign */
+        matchingUser.saml = {};
+        /* eslint-disable-next-line no-param-reassign */
+        matchingUser.saml.nameID = profile.nameID;
+        /* eslint-disable-next-line no-param-reassign */
+        matchingUser.saml.nameIDFormat = profile.nameIDFormat;
+        done(
+          matchingUser ? null : new Error('Email is not registered: ' + email),
+          matchingUser
+        );
+      });
+  }
+);
+
+passport.use(samlStrategy);
+
+/** This saves user.email in the session, under req.session.passport.user,
+ *    which is later used by deserializeUser to find the user object and attach 
+ *    it to req.user. After this operation, req.session.passport.user = user.email */ 
+passport.serializeUser((user, done) => {
+  done(null, user.email);
+});
+
+
+/** This is called with the email that was saved in serializeUser as the
+*     first argument to this function. (i.e. email = user.email). We would then
+*     use the user's email to get the actual user object (using db.user.findOne(...))
+*     Afterwards, the user object is attached to req.user */
```

Comment:
what's the difference between `req.session.passport.user` vs `req.user`?

Reply:
so `req.session.passport.user` is used by passport to deserialize the user object, while `req.user` is the deserialized object. I'm guessing that `req.session.passport.user` includes extraneous metadata for `passport`.